utils/src/main/java/com/booking/validator/utils/ConcurrentPipeline.java

```diff
@@ -48,12 +48,17 @@ public void stop(){
     }
 
     private void startTaskAsync(){
-
         if (run) CompletableFuture.supplyAsync(supplier)
                 .thenCompose( Supplier::get )
                 .whenComplete( consumer )
-                .whenComplete( (x,t)->startTaskAsync()); // a consumer exception could be handled here
-
+                .whenComplete( (x,t)-> {
+                    try {
+                        Thread.sleep(100);
```

Comment:
This callback is being executed on a shared fork-join pool which by default is limited. Its threads should not be blocked especially by not doing any job. I would propose at to move the pipe from a default thread pool to an owned at least. The whole concept of "saving cpu" looks unclear: before the changes the validator was controlled by back pressure of the results sink/data-providers. Did it change?

Reply:
This sleep should definitely be removed from the shared fork-join pool. I will implement a separate thread pool and move the rate limiting to the supplier as suggested.